### PR TITLE
Support musl builds along with packaging

### PR DESCRIPTION
**Pull request by anupdhml**
Package formats supported currently are archive (tar.gz) and deb.

Some documentation at: https://github.com/wayfair-tremor/tremor-runtime/tree/musl_plus_packaging/packaging

### CODE REVIEW

**Comment by Licenser** [comment]
Currenty the build fails for musl with

```
error[E0308]: mismatched types
   --> /cargo/registry/src/github.com-1ecc6299db9ec823/simd-json-0.3.9/src/lib.rs:216:86
    |
216 | fn please_compile_with_a_simd_compatible_cpu_setting_read_the_simdjonsrs_readme() -> ! {}
    |    ----------------------------------------------------------------------------      ^ expected `!`, found `()`
    |    |
    |    implicitly returns `()` as its body has no tail or `return` expression
    |
    = note:   expected type `!`
            found unit type `()`
```

Which is a bit surprising since based on the output the flags are set correctly.

File: packaging/cross_build.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+###############################################################################
+
+# install cross if not already there (helps us build easily across various targets)
+# see https://github.com/rust-embedded/cross
+#
+# currently need to install it from a personal fork for builds to work against custom targets
+# (eg: x86_64-alpine-linux-musl which we use for generating working musl binaries right now)
+if ! command -v cross > /dev/null 2>&1; then
+  echo "Installing cross..."
+  cargo install --git https://github.com/anupdhml/cross.git --branch custom_target_fixes
```
Comment (Licenser) [inline packaging/cross_build.sh:51]: Can we upstream that?

File: packaging/cross_build.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  CUSTOM_RUSTFLAGS+=("-C" "target-cpu=native")
+fi
+
+if [[ "$TARGET" == *"alpine-linux-musl"* ]]; then
+  # force static binaries for alpine-linux-musl targets (since we are choosing this
+  # target specifically to produce working static musl binaries). Static building
+  # is the default rustc behavior for musl targets, but alpine disables it by
+  # default (via patches to rust).
+  echo "Ensuring static builds for alpine-linux-musl targets..."
+  CUSTOM_RUSTFLAGS+=("-C" "target-feature=+crt-static")
```
Comment (Licenser) [inline packaging/cross_build.sh:81]: This overwrites the target-features set before
```suggestion
  CUSTOM_RUSTFLAGS+=("-C" "target-feature=+crt-static,+avx,+avx2,+sse4.2")
```

Comment (anupdhml) [reply packaging/cross_build.sh]: good catch!

File: packaging/cross_build.sh (hunk 1 above)
Comment (anupdhml) [reply packaging/cross_build.sh:51]: will do. I wanted to try out some other things on that branch before filing a PR -- will test them out soon and submit it upstream.

File: packaging/cross_build.sh (hunk 2 above)
Comment (Licenser) [reply packaging/cross_build.sh]: 👍 I like the storing it as list approach a lot better :D great ide!

File: packaging/cross_build.sh (hunk 1 above)
Comment (anupdhml) [reply packaging/cross_build.sh:51]: got some time today to finally make that upstream PR!

https://github.com/rust-embedded/cross/pull/431

File: packaging/builder-images/Dockerfile.x86_64-unknown-linux-musl [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    DL_CMD="curl -C - -L -o" \
+    OUTPUT=/usr/local/ \
+    TARGET=x86_64-linux-musl \
+  && rm -rf $temp_dir
+
+ENV CARGO_TARGET_X86_64_UNKNOWN_LINUX_MUSL_LINKER=x86_64-linux-musl-gcc \
+    CC_x86_64_unknown_linux_musl=x86_64-linux-musl-gcc \
+    CXX_x86_64_unknown_linux_musl=x86_64-linux-musl-g++
+
+RUN rustup target add x86_64-unknown-linux-musl
```
Comment (anupdhml) [inline packaging/builder-images/Dockerfile.x86_64-unknown-linux-musl:26]: filed an issue on cross upstream to see if we can bump up the gcc version on the default cross images:
https://github.com/rust-embedded/cross/issues/432

If we can do that, we can derive this image from ` rustembedded/cross:x86_64-unknown-linux-musl` and not have to install musl ourselves.

File: rust-toolchain [hunk 4]
```diff
@@ -0,0 +1 @@
+1.43.1
```
Comment (Licenser) [inline rust-toolchain:1]: Got a new rust version :D
```suggestion
1.44.0
```

**Pull request merged**